packages/actor-query-operation-from-quad/lib/ActorQueryOperationFromQuad.ts

```diff
@@ -16,58 +16,124 @@ export class ActorQueryOperationFromQuad extends ActorQueryOperationTypedMediate
-        return defaultGraph ? ActorQueryOperationFromQuad.unionOperations(graphs.map(
-          (graph: RDF.Term) => ActorQueryOperationFromQuad.applyOperationGraph(operation, [graph])))
+        return defaultGraph ? ActorQueryOperationFromQuad.unionOperations(defaultGraphs.map(
+          (graph: RDF.Term) => ActorQueryOperationFromQuad.applyOperationDefaultGraph(operation, [graph])))
           : operation;
       }
     }
 
-    // Otherwise, copy the operation and recursively call this method.
-    const copiedOperation: Algebra.Operation = <any> {};
-    for (const key of Object.keys(operation)) {
-      if (Array.isArray(operation[key])) {
-        if (key === 'variables') {
-          copiedOperation[key] = operation[key];
+    return ActorQueryOperationFromQuad.copyOperation(operation,
+      (subOperation: Algebra.Operation) => this.applyOperationDefaultGraph(subOperation, defaultGraphs));
+  }
+
+  /**
+   * Recursively transform the given operation to use the given graphs as named graph
+   * This will (possibly) create a new operation and not modify the given operation.
+   * @param {Operation} operation An operation.
+   * @param {RDF.Term[]} namedGraphs Graph terms.
+   * @return {Operation} A new operation.
+   */
+  public static applyOperationNamedGraph(operation: Algebra.Operation, namedGraphs: RDF.Term[],
+                                         defaultGraphs: RDF.Term[]): Algebra.Operation {
+    // If the operation is a Pattern or Path, change the graph.
+    if (operation.type === 'pattern' || operation.type === 'path') {
+      const patternGraph: RDF.Term = operation.graph;
+      if (patternGraph.termType === 'DefaultGraph') {
+        // SPARQL spec (8.2) describes that when FROM NAMED's are used without a FROM, the default graph must be empty.
+        // The FROMs are transformed before this step to a named node, so this will not apply to this case anymore.
+        return { type: 'bgp', patterns: [] };
+      } else if (patternGraph.termType === 'Variable') {
+        if (namedGraphs.length === 1) {
+          const graph: RDF.Term = namedGraphs[0];
+          // If the pattern graph is a variable, replace the graph and bind the variable using VALUES
+          const bindings: {[key: string]: RDF.Term} = {};
+          bindings['?' + patternGraph.value] = graph;
+          const values: Algebra.Values = ActorQueryOperationFromQuad.FACTORY
+            .createValues([<RDF.Variable> patternGraph], [bindings]);
+          let pattern: Algebra.Operation;
+          if (operation.type === 'pattern') {
+            pattern = ActorQueryOperationFromQuad.FACTORY
+              .createPattern(operation.subject, operation.predicate, operation.object, graph);
+          } else {
+            pattern = ActorQueryOperationFromQuad.FACTORY
+              .createPath(operation.subject, operation.predicate, operation.object, graph);
+          }
+          return ActorQueryOperationFromQuad.FACTORY.createJoin(values, pattern);
```

Comment:
Not sure what you mean, but I assume this was a misunderstanding.

This method rewrites single patterns or paths. If a BGP is found, then every pattern in this BGP is transformed like this.

Reply:
Assume you have a query with a bgp (as tends to happen), e.g.: `{ type: bgp, patterns: [QUAD]}`, with QUAD having a variable graph. This means a VALUES block gets added with a JOIN. Does this result in `{ type: bgp, patterns: [ {type: join, left: QUAD, right: VALUES}] }` or in `{ type: join, left: { type: bgp, patterns: [QUAD] }, right: VALUES }`? Because the first one is incorrect (bgp only allows quads in the pattern array, not any operation). (And in the second one the bgp would have to be split into multiple bgps if there are multiple patterns in its array).

Or I am misreading what's going on in the code :D

Reply:
You're right, made an issue for it #105.